Log items that share the same UUID

Spazcool opened the pull request:
REFACTOR: logScanned to catch doubled UUIDs & render doubles on template

Commit:
REFACTOR: logScanned to catch doubled UUIDs & render doubles on template

lenzai commented:
![image](https://user-images.githubusercontent.com/2392519/34081051-7fcf56ba-e382-11e7-967a-e58fe603d28f.png)

I managed to crash the "page not found" in this branch.

lenzai commented:
The code both in the logScanned and template introduces duplication.

lenzai reviewed (changes requested):
there s a bug

cannot accept code duplication

Commit:
FIX: hacked out bug

Commit:
REMOVE: cut duplication in logScanned func

Commit:
REFACTOR: logScanned works again

Commit:
REMOVE: double UUID conditional from recent.ejs

Commit:
ADD: double UUID conditional without duplicate code

Commit:
ADD: double UUID conditional without duplicate code

Commit:
ADD: link to item row in spreadsheet to both item obj and recent.ejs

Spazcool commented:
@lenzai should be better now. I removed that bug you found as well as the duplicated code. I also added links to item's row in the spreadsheet, should an edit be desired.

lenzai commented on app.js:
```diff
@@ -44,9 +40,10 @@ app.get(['/favicon.ico', '/robots.txt'], (req, res) => {
-        .sort((a, b) => (a.floor === b.floor ? 0 : +(a.floor > b.floor) || -1)),
+      matches: searchDatabase(req.query, allItems).sort((a, b) =>
+        (a.floor === b.floor ? 0 : +(a.floor > b.floor) || -1)),
     });
```
this is noise, refrain from doing cosmetic changes.

mpsido replied:
you may understand why if you try the git blame command

lenzai commented on googleSpreadsheet.js:
```diff
@@ -1,11 +1,12 @@
     o[lab[i]] = val[i];
   }
+  o.cellRef = 'https://docs.google.com/spreadsheets/d/1QHKa3vUpht7zRl_LEzl3BlUbolz3ZiL8yKHzdBL42dY/edit#gid=0&range=A' + index + ':T' + index;
   return o;
```
potential security issue in leaking spreadsheet link to front end

hard coded value is pretty bad.

mpsido replied:
agree, why do we need a cellRef here ?

Spazcool replied:
Not a necessary addition but, I thought having a link to any troublesome item's place in the spreadsheet would be useful for whoever has to edit it.

mpsido replied:
that is a fair concern, and actually you can argue that google spreadsheet retrains access rights to the priviledged users. 

on further development we may need a better strategy for that spreadsheet bug fixing.

All that said I still approve that pull request

lenzai commented on app.js:
```diff
@@ -8,18 +8,26 @@ const app = express();
-    return;
-  }
+function logScanned(uuid, matches) {
+  let allMatches = matches;
```
redundant

Spazcool replied:
This in particular was to avoid linter errors.

lenzai reviewed (commented):
seems that we are in rabbit hole of moving around more and more information. (i.e. increasing coupling and thus technical debt) in order to just print the perfect logging message.

Logging is supposed to help debugging the application.
Increasing complexity sooner or later will create bugs or at least increase maintenance cost.

mpsido commented on views/recent.ejs:
```diff
@@ -1,26 +1,25 @@
+        : <br/>
+        <a href="/<%- allScans[i].uuid %>"><%- allScans[i].uuid %></a>
+        <% if (!allScans[i].fixture) { %>
+          <a href="<%- allScans[i].link %>" style="color: <%- allScans[i].status === 'fixed' ? 'orange' : 'red' %>">Missing Details!</a>
```
it makes sense to edit style inline here because all we want is to change a color.
but it makes the job of a designer who only knows css more difficult.
consider just giving a style class (class=fixed or class=broken) and edit that style in the css file

mpsido commented:
great job!

Spazcool commented:
@lenzai I thought the logging here was for whoever is editing the spreadsheets not whoever is debugging the code, even if both groups can use the logs.

mpsido committed:
Refactoring/simplifying code googleSpreadsheet.js

mpsido committed:
Removing hard url in app.js

lenzai commented:
@Spazcool, yes it's for the content editors mostly which needs to "debug" data.
Requirements and enhancement you defined are really thoughtful. it's just that the implementation need to be simplified.

Some reference concept:
https://en.wikipedia.org/wiki/Coupling_(computer_programming)
https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)

It's quite difficult to discuss remotely, without either dictating you step by step what to do or just publishing revised code myself. If above links inspire you, feel free to discuss or commit further. Otherwise, we can have a pair programming session in January.

PS: So far, most of the code reviews by @mpsido were published through a mobile phone... I think I fixed that today.

mpsido committed:
Refactoring, addRecentlyScanned function instead of logScanned

mpsido commented:
Pushed some simplifications on that code

lenzai reviewed (commented):
not a good practice to mix refactoring and altering behaviour ( i.e. deleting cellRef)

renaming is thoughtful

introducing redundant local variable is not recommended. It can be a symptom of poor function breakdown

lenzai commented on app.js:
```diff
@@ -6,28 +6,26 @@ const { loadDatabase, searchDatabase } = require('./googleSpreadsheet');
-    allMatches[0].double = true;
+
+  const duplicatedItem = item;
+
```
`duplicatedItem` is redundant with `item` we don't need 2 names for 1 thing.

mpsido replied:
I don't know if item is a reference or a copy.
I wanted to ensure it is a copy because I am writing in it.

lenzai replied:
I don't think this is creating any copy.

mpsido replied:
I need to recheck that, but if that is a reference, then I need the local variable

lenzai replied:
not really useful reply

lenzai commented on views/recent.ejs:
```diff
@@ -2,21 +2,21 @@
   <ul class='recent-list'>
-    <% for (var i = 0; i < allScans.length; i++) { %>
+    <% for (let item of allScans.values()) { %>
       <li>
```
@mpsido nice! Are there other areas in the code where this rewrite should be applied?

mpsido replied:
Everywhere there is a for loop ;) if you don't care about the index then you can use for..in or for..of.
See the MSDN documentation to understand the difference ;)

lenzai replied:
I mean do it.

lenzai commented on app.js:
```diff
@@ -6,28 +6,26 @@ const { loadDatabase, searchDatabase } = require('./googleSpreadsheet');
 
-const allScans = [];
+const allScans = new Map();
 
```
why?

mpsido replied:
Because I want to avoid repetitions in the "recent" page.
I want to add one by one unique items on that page

lenzai replied:
This is not the way to handle changes and features. We create an issue, discuss it and push it.
Hidding secret features in unrelated commits is not appropriate decision workflow and distracting for code review

lenzai commented on googleSpreadsheet.js:
```diff
@@ -1,13 +1,14 @@
-  o.cellRef = 'https://docs.google.com/spreadsheets/d/1QHKa3vUpht7zRl_LEzl3BlUbolz3ZiL8yKHzdBL42dY/edit#gid=0&range=A' + index + ':T' + index;
-  return o;
+  return formatedRow;
 }
```
and now the template for `http:///recent` display irrelevant links to home page.

lenzai reviewed (changes requested):
some steps towards less complexity ( improved encapsulation and naming)
some off topic changes
introducing multiple bugs

see code comments for details

Spazcool commented on app.js:
```diff
@@ -75,6 +75,9 @@ app.get('/:uuid', (req, res) => {
+    if (matches.length >= 1) {
+      console.log(`Too much matches for uuid ${req.params.uuid} length = ${matches.length}`);
+    }
     matches[0].similarItems = searchDatabase({ fixture: matches[0].fixture }, allItems)
```
Perhaps amateur questions, but why remove the hard URL and how else do we get a link to work?

mpsido commented on app.js:
```diff
@@ -54,7 +54,7 @@ app.get('/qrlist', (req, res) => {
       .sort((a, b) => (a.floor === b.floor ? 0 : +(a.floor > b.floor) || -1));
-    qrList.forEach(item => item.qr = qr.imageSync('http://url.coderbunker.com/' + item.uuid, { type: 'svg' }));
+    qrList.forEach(item => item.qr = qr.imageSync(item.uuid, { type: 'svg' }));
     res.render('qrList', { matches: qrList });
```
I'll fix that, in which case it won't work ? I need to reproduce it so that I ensure the change will fix that case

lenzai replied:
if current page is not root folder then it will assume the uuid is a subfolder

mpsido commented:
@lenzai what are the introduced bugs you are talking about ?

mpsido commented on googleSpreadsheet.js:
```diff
@@ -21,8 +22,10 @@ function loadDatabase(callback) {
+    // map function transforms a list into another list using the given lambda function
+    const formatedRows = response.values.map(row => spreadsheetValuesToObject(row, columns));
+    return callback(formatedRows);
   });
```
I actually meant to break into two steps for better readability,
All the things where nested into one line of code that was a headache to read

lenzai replied:
Extracting local variable is usually bad design. 
Extracting subfunction is fine (while inlining others maybe)

mpsido replied:
I don't understand what "Extracting subfunction" means.
If javascript is smart (and I assume it is), formatedRows is passed by reference.
It does improve readbility don't you think so ?

lenzai replied:
- extracting sub function https://refactoring.com/catalog/extractMethod.html
- removing local variables https://refactoring.com/catalog/inlineTemp.html

mpsido commented:
Closing this pull request
see that one instead:
https://github.com/coderbunker/inventory-server/pull/62

mpsido committed:
Merge master

mpsido committed:
Fixing pre-commit lint warnings, and adding addQrImg function to items

mpsido committed:
moving filterEmptyUuid and sortByFloor functions to attribute of matches list

mpsido committed:
Removing deprecated similar items function

mpsido committed:
Fixing qrList regression

mpsido committed:
Reintroducing cellRef

mpsido commented:
shall we merge it and close the associated issue ?

mpsido commented:
@lenzai what do you think we should do with this PR
